afpacket/src/sockets.rs

```diff
@@ -0,0 +1,181 @@
+#![deny(missing_docs)]
+
+use libc;
+use std::{
+    io,
+    ptr,
+    mem::{self, MaybeUninit},
+    ffi::CStr};
+use crate::linux;
+
+/// Represents a link-local address.
+/// At this time, it's not particularly useful.
+pub struct Addr {
+    inner: libc::sockaddr_storage,
+    len: libc::socklen_t,
+}
+
+/// Represents an unbound `AF_PACKET` socket.  At this phase of a socket's lifecycle, it can be
+/// configured.
+pub struct Socket {
+    fd: libc::c_int,
+}
+
+/// Represents a bound `AF_PACKET` socket. At this phase of a socket's lifecycle, it can be read
+/// to/written from.
+pub struct BoundSocket {
+    fd: libc::c_int,
+    send_addr: libc::sockaddr_ll,
+}
+
+impl Socket {
+    /// Creates a new unbound socket.
+    pub fn new() -> io::Result<Self> {
+        // This block is unsafe because it invokes a syscall. It is safe because it checks the
+        // syscall return value and returns an Err if the sycall fails. It does not interact with
+        // Rust-owned memory, so it's safe as per the borrow checker's rules.
+        let fd = unsafe {
+            // Resources:
+            // https://beej.us/guide/bgnet/html/multi/syscalls.html#socket
+            // man 7 packet
+            let fd = libc::socket(libc::AF_PACKET, libc::SOCK_RAW, libc::ETH_P_ALL.to_be());
```

Comment:
"to be or not to be"

Reply:
Why are we manipulating endianness of `ETH_P_ALL`? I also couldn't really figure out what that meant. Is it "give me all the packets regardless of packet type"? [Relevant search](https://docs.rs/libc/0.2.54/libc/constant.ETH_P_ALL.html?search=eth_p)

Reply:
> When protocol is set to htons(ETH_P_ALL), then all protocols are received.  All incoming packets of that protocol type will be passed to the packet socket before they are passed to the protocols implemented in the kernel.

[source](http://man7.org/linux/man-pages/man7/packet.7.html).

`htons` in C is "host to network byte order for `short`" (i.e. convert unsigned 16-bit integer to big-endian, which is network byte order). In Rust, that's `to_be`.

Reply:
Neat! TIL "network byte order" is big-endian.